api/src/main/java/jakarta/json/bind/customization/JsonbCustomizationBuilder.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package jakarta.json.bind.customization;
+
+import java.text.DateFormat;
+import java.text.NumberFormat;
+import java.util.Locale;
+
+import jakarta.json.bind.adapter.JsonbAdapter;
+import jakarta.json.bind.serializer.JsonbDeserializer;
+
+/**
+ * Common interface for all customization builders.
+ */
+public interface JsonbCustomizationBuilder<T extends JsonbCustomizationBuilder<T, B>, B extends JsonbCustomization> {
+
+    /**
+     * Whether this component can be nillable.
+     *
+     * @param nillable nillable component
+     * @return updated builder instance
+     */
+    T nillable(boolean nillable);
+
+    /**
+     * Ignore nillable customization of this component.
+     *
+     * @return updated builder instance
+     */
+    T ignoreNillable();
+
+    /**
+     * Set {@link JsonbDeserializer} which should be used.
+     *
+     * @param deserializer component deserializer
+     * @return updated builder instance
+     */
+    T deserializer(JsonbDeserializer<?> deserializer);
+
+    /**
+     * Ignore deserializer of this component.
```

Comment:
`ignoreNillable` above says "Ignore nillable *customization*", while this says "Ignore deserializer".

What is the semantic of these ignore methods? Resetting the customization, or overriding default beaviour, e. g. ignoring deserializer when one is defined with an annotation?

Reply:
The whole concept of the ignore methods should be in place to ignore specific annotation on the component, other customizations (and also inherited ones) which apply to the component should be applied. I am still playing a bit with an idea of having just a single `ignore(Class<? extends Annotation>)` method. I am still not sure if that is not more clear.

Anyway yes, javadoc of this version is not very clear. I do have it fixed locally, but not yet pushed :-)

Reply:
The reason I asked was because one can also do `somePropertyCustomization.toBuilder().ignoreDeserializer()`. What is the effect of this call if `somePropertyCustomization` defined a serializer?

The customization of every attribute is from domain `Default|None|Custom(value)`, where Default represents no customization, so whatever spec+class annotation prescribe, None represents spec while ignoring class annotation, and finally Custom is customized value from customizer. Usually one starts at Default and with `ignore` method can set processing to `None`, with `set(x)` method can set processing to `Custom(x)`. But with `toBuilder` you can start in any of the states and you need transition path to any other state. The common ignore method will only solve part of the problem, while introducing the binding to annotations and it is clear where it would go from there ;)

Clean model of this would be to have every customization attribute represented as type similar to `Optional<T>`, where it is either `Customization.ignored()`, `Customization.annotated()` or `Customization<T>.of(T value)`. It is possible to have single getter/setter pair for every attribute and cover all of the possible states.

Additionally a convenience setter can be added, where non-null value represents `of(value)` and null represents `annotated()`.

Another approach is to have provider to return Builders with full mapping, but that turns it into introspection API.

Reply:
This also shows weakness of `toBuilder()` methods. It cannot be really reused for different property names, in case I have common customization to apply to more fields. Either names should also be modifiable in builders or method `toBuilder(name)` should also exist.

One final alternative is to drop `toBuilder()` methods, and rely on `TypeCustomizationBuilder.property(String,Consumer)` for sharing settings. Then everything starts in `annotated` state.